pkg/retriever/bitswapretriever.go

```diff
@@ -187,57 +184,66 @@ func (br *bitswapRetrieval) RetrieveFromAsyncCandidates(ayncCandidates types.Inb
 	// loader that can work with bitswap just for this session
 	loader := loaderForSession(br.request.RetrievalID, br.inProgressCids, br.bsGetter)
 
-	// set up the storage system, including the preloader
-	concurrency := br.BitswapRetriever.cfg.Concurrency
-	if concurrency <= 0 {
-		concurrency = DefaultConcurrency
-	}
-	cacheLinkSys := br.request.PreloadLinkSystem
-	if cacheLinkSys.StorageReadOpener == nil || cacheLinkSys.StorageWriteOpener == nil {
-		// An uninitialised LinkSystem, i.e. likely the request didn't provide a
-		// PreloadLinkSystem. So we'll create a new one. This will result in a new
-		// temporary file being created for the retrieval.
-		tmpDir := br.BitswapRetriever.cfg.TempDir
-		if tmpDir == "" {
-			tmpDir = os.TempDir()
+	// set up the storage system, including the preloader if configured
+	var preloader preload.Loader
+	traversalLinkSys := br.request.LinkSystem
+	var storage *bitswaphelpers.PreloadCachingStorage
+
+	if br.request.PreloadLinkSystem.StorageReadOpener != nil && br.request.PreloadLinkSystem.StorageWriteOpener != nil {
+		concurrency := br.BitswapRetriever.cfg.Concurrency
+		if concurrency <= 0 {
+			concurrency = DefaultConcurrency
 		}
-		cacheStore := storage.NewDeferredStorageCar(tmpDir)
-		cacheLinkSys = cidlink.DefaultLinkSystem()
-		cacheLinkSys.SetReadStorage(cacheStore)
-		cacheLinkSys.SetWriteStorage(cacheStore)
-		cacheLinkSys.TrustedStorage = true
-		defer cacheStore.Close()
-	}
-	storage, err := bitswaphelpers.NewPreloadCachingStorage(
-		br.request.LinkSystem,
-		cacheLinkSys,
-		loader,
-		concurrency,
-	)
-	if err == nil {
-		err = storage.Start(ctx)
-	}
-	if err != nil {
-		cancel()
-		return nil, err
-	}
 
-	// setup providers linksystem for this retrieval
-	br.bstore.AddLinkSystem(
-		br.request.RetrievalID,
-		bitswaphelpers.NewByteCountingLinkSystem(storage.BitswapLinkSystem, cb),
-	)
+		var err error
+		storage, err = bitswaphelpers.NewPreloadCachingStorage(
+			br.request.LinkSystem,
+			br.request.PreloadLinkSystem,
+			loader,
+			concurrency,
+		)
+		if err == nil {
+			err = storage.Start(ctx)
+		}
+		if err != nil {
+			cancel()
+			return nil, err
+		}
+		preloader = storage.Preloader
+		traversalLinkSys = *storage.TraversalLinkSystem
+
+		br.bstore.AddLinkSystem(
+			br.request.RetrievalID,
+			bitswaphelpers.NewByteCountingLinkSystem(storage.BitswapLinkSystem, bytesWrittenCb),
+		)
+	} else {
+		br.bstore.AddLinkSystem(
+			br.request.RetrievalID,
+			bitswaphelpers.NewByteCountingLinkSystem(&br.request.LinkSystem, bytesWrittenCb),
+		)
+		traversalLinkSys.StorageReadOpener = func(lc linking.LinkContext, l datamodel.Link) (io.Reader, error) {
```

Comment:
This isn't neccesary I believe. The BlockService will attempt a local load before going to Bitswap.

Reply:
If I remove this, these tests fail:

* `TestDirectFetch/direct_bitswap_peer`
* `TestBitswapRetriever/without_preloader/successful_full_remote_fetch`
* `TestBitswapRetriever/without_preloader/fail_remote_fetch_about_non-zero_blocks`
* `TestBitswapRetriever/without_preloader/successful_selective_remote_fetch`
* `TestBitswapRetriever/without_preloader/timeout`
* `TestBitswapRetriever/without_preloader/successful_partial_remote_fetch`

Although their failures don't have the usual "everything is busted" look, I'm suspecting this might be just about the root block, which we fetch ourselves in `easyTraverse`. Will investigate further.